compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java

```diff
@@ -544,64 +586,102 @@ private void checkUniqueness(BLangConstant constant) {
 
     private void updateConstantType(BLangConstant constant) {
         BConstantSymbol symbol = constant.symbol;
-        if (symbol.type.getKind() == TypeKind.FINITE) {
+        updateConstantType(symbol, constant.expr, symEnv);
+
+        BType resolvedType = symbol.type;
+
+        if (resolvedType.tag == TypeTags.INTERSECTION) {
+            BIntersectionType intersectionType = (BIntersectionType) resolvedType;
+
+            if (intersectionType.effectiveType.tag == TypeTags.RECORD) {
+                addAssociatedTypeDefinition(constant, intersectionType);
+            }
+        }
+    }
+
+    private void updateConstantType(BConstantSymbol symbol, BLangExpression expr, SymbolEnv env) {
+        BType type = Types.getReferredType(symbol.type);
+        if (type.getKind() == TypeKind.FINITE) {
             return;
         }
 
         if (symbol.value == null) {
             return;
         }
 
-        BType resolvedType = checkType(constant.expr, constant, symbol.value.value, symbol.type, symbol.pos,
-                                        constant.symbol.value.value);
-        if (resolvedType != null) {
-            if (constant.symbol.type.getKind() == TypeKind.MAP && resolvedType.getKind() == TypeKind.INTERSECTION) {
-                addassociatedTypeDefinition(constant, (BIntersectionType) resolvedType);
-                constant.symbol.type = resolvedType;
-                constant.expr.setBType(((BIntersectionType) resolvedType).effectiveType);
-                constant.symbol.literalType = resolvedType;
-                symbol.value.type = resolvedType;
-                return;
-            }
-            constant.expr.setBType(resolvedType);
-            constant.symbol.type = resolvedType;
+        BType resolvedType = checkType(expr, symbol, symbol.value.value, type, symbol.pos, env);
+        if (resolvedType == null) {
+            return;
+        }
+
+        if (isListOrMapping(type.tag) && resolvedType.getKind() == TypeKind.INTERSECTION) {
```

Comment:
shall we swap condition order here. We need to check `isListOrMapping` only if the resolution is a intersection type.